### PR TITLE
Reserved Cores [4/4]: Implement driver cpuset cgroup path consumption

### PR DESCRIPTION
**Pull request by nickethier**
This PR implements the driver side of reserved cores.

The `exec`, `java` and `qemu` drivers all use the executor to launch processes. The Linux executor flavor has been changed to add a [hook](https://github.com/opencontainers/runtime-spec/blob/1c3f411f041711bbeecf35ff7e93461ea6789220/config.md#posix-platform-hooks) which switches the cpuset cgroup of the init process prior to switching to the user process. The reason being that the libcontainer cgroup manager configuration is all or nothing. I didn't want to manually manage the cpu and memory cgroups. This seemed like the best compromise as the cgroup switch still happens prior to the user process starting.

For the `docker` driver, there is a single `Cgroup` config option which sets the group path for all subsystems. For this reason the only option was to switch the cpuset cgroup immediately after starting the container. This means there is a very brief window where the cpuset will include all system cpus before switching the the shared or reserve sets dictated by Nomad.

The `raw_exec` driver will not support cpuset configuration as with other resource isolation systems.

### CODE REVIEW

File: plugins/drivers/proto/driver.proto [hunk 1]
```diff
@@ -525,10 +525,9 @@ message LinuxResources {
     int64 memory_limit_bytes = 4;
     // OOMScoreAdj adjusts the oom-killer score. Default: 0 (not specified)
     int64 oom_score_adj = 5;
-    // CpusetCpus constrains the allowed set of logical CPUs. Default: "" (not specified)
-    string cpuset_cpus = 6;
-    // CpusetMems constrains the allowed set of memory nodes. Default: "" (not specified)
-    string cpuset_mems = 7;
```
Comment (tgross) [inline plugins/drivers/proto/driver.proto:531]: I don't think we can safely remove fields we've [already shipped](https://github.com/hashicorp/nomad/blob/v1.0.4/plugins/drivers/proto/driver.proto#L527-L530)... we own the drivers that we ship, but the community drivers will all be broken.

File: drivers/shared/executor/executor_linux.go [hunk 2]
```diff
@@ -884,3 +895,9 @@ func lookPathIn(path string, root string, bin string) (string, error) {
 	}
 	return "", fmt.Errorf("file %s not found under path %s", bin, root)
 }
+
+func newSetCPuSetCgroupHook(cgroupPath string) lconfigs.Hook {
```
Comment (tgross) [inline drivers/shared/executor/executor_linux.go:899]: This probably should be `newSetCpuSetCgroupHook` or `newSetCPUSetCgroupHook`.

**Review by tgross — CHANGES_REQUESTED** [review]
The new code here LGTM but I don't think we can remove fields from the protobufs without breaking downstream drivers in the community (ex. podman or Roblox's containerd)

File: plugins/drivers/proto/driver.proto (hunk 1 above)
Comment (nickethier) [reply plugins/drivers/proto/driver.proto:531]: These fields are never used, never even set as far as I could tell. I'm ok leaving them in, but the Nomad client never sets a value for them (as far as I can tell).

Comment (tgross) [reply plugins/drivers/proto/driver.proto:531]: I think we're supposed to "reserve" the existing field numbers rather than reuse them in that case.

Comment (nickethier) [reply plugins/drivers/proto/driver.proto:531]: It looks like this breaks our linter anyways (for good reason) so I'll just revert and make a new field:
```
==> Checking for breaking changes in protos...
plugins/drivers/proto/driver.proto:516:1:Previously present field "7" with name "cpuset_mems" on message "LinuxResources" was deleted without reserving the name "cpuset_mems".
plugins/drivers/proto/driver.proto:516:1:Previously present field "7" with name "cpuset_mems" on message "LinuxResources" was deleted without reserving the number "7".
plugins/drivers/proto/driver.proto:529:5:Field "6" with name "cpuset_cgroup" on message "LinuxResources" changed option "json_name" from "cpusetCpus" to "cpusetCgroup".
plugins/drivers/proto/driver.proto:529:12:Field "6" on message "LinuxResources" changed name from "cpuset_cpus" to "cpuset_cgroup".
make: *** [GNUmakefile:137: check] Error 1
```

**Review by tgross — APPROVED** [review]
LGTM! 👍

**Pull request merged**